homeassistant/components/mysensors/config_flow.py

```diff
@@ -0,0 +1,229 @@
+"""Config flow for MySensors."""
+import logging
+from typing import Dict, Optional
+
+import voluptuous as vol
+
+from homeassistant.components.mysensors import (
+    CONF_DEVICE,
+    DEFAULT_BAUD_RATE,
+    DEFAULT_TCP_PORT,
+    is_persistence_file,
+)
+import homeassistant.helpers.config_validation as cv
+
+from . import CONF_RETAIN, CONF_VERSION, DEFAULT_VERSION
+from ... import config_entries
+from ..mqtt import valid_publish_topic, valid_subscribe_topic
+from .const import (
+    CONF_BAUD_RATE,
+    CONF_GATEWAY_TYPE,
+    CONF_GATEWAY_TYPE_ALL,
+    CONF_GATEWAY_TYPE_MQTT,
+    CONF_GATEWAY_TYPE_SERIAL,
+    CONF_GATEWAY_TYPE_TCP,
+    CONF_GATEWAY_TYPE_TYPE,
+    CONF_PERSISTENCE_FILE,
+    CONF_TCP_PORT,
+    CONF_TOPIC_IN_PREFIX,
+    CONF_TOPIC_OUT_PREFIX,
+    DOMAIN,
+)
+from .gateway import MQTT_COMPONENT, is_serial_port, is_socket_address, try_connect
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _get_schema_common() -> dict:
+    """Create a schema with options common to all gateway types."""
+    schema = {
+        vol.Required(
+            CONF_VERSION, default="", description={"suggested_value": DEFAULT_VERSION}
+        ): str,
+        vol.Optional(
+            CONF_PERSISTENCE_FILE,
+        ): str,
+    }
+    return schema
+
+
+def _validate_version(version: str) -> Dict[str, str]:
+    """Validate a version string from the user."""
+    errors = {CONF_VERSION: "invalid_version"}
+    version_parts = version.split(".")
+    if len(version_parts) != 2:
+        return errors
+    try:
+        [int(x) for x in version_parts]
+    except ValueError:
+        return errors
+    return {}
+
+
+class MySensorsConfigFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow."""
+
+    async def async_step_import(self, user_input: Optional[Dict[str, str]] = None):
+        """Import a config entry.
+
+        This method is called by async_setup and it has already
+        prepared the dict to be compatible with what a user would have
+        entered from the frontend.
+        Therefore we process it as though it came from the frontend.
+        """
+        if user_input.get(CONF_DEVICE) == MQTT_COMPONENT:
+            user_input[CONF_GATEWAY_TYPE] = CONF_GATEWAY_TYPE_MQTT
+        else:
+            try:
+                await self.hass.async_add_executor_job(
+                    is_serial_port, user_input[CONF_DEVICE]
+                )
+            except vol.Invalid:
+                user_input[CONF_GATEWAY_TYPE] = CONF_GATEWAY_TYPE_TCP
+            else:
+                user_input[CONF_GATEWAY_TYPE] = CONF_GATEWAY_TYPE_SERIAL
+
+        return await self.async_step_user(user_input=user_input)
+
+    async def async_step_user(self, user_input: Optional[Dict[str, str]] = None):
+        """Create a config entry from frontend user input."""
+        schema = dict()
```

Comment:
Please create dicts with the literal expression `{}`. We can populate the dict directly.